File: AStar/2024baiduStar/2024baiduStar01/B.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
const int N = 1e7 + 5;
const ll mod = 998244353;
bool isprime[N];
int prime[N];
int cnt;
int minp[N];
ll inv[N];
int n;
void init(int n)
{
    memset(isprime, 1, sizeof isprime);
    isprime[1] = false;
    for (int i = 2; i <= n; i++)
    {
        if (isprime[i])
        {
            prime[++cnt] = i;
            minp[i] = i;
        }
        for (int j = 1; j <= cnt && i * prime[j] <= n; j++)
        {
            isprime[i * prime[j]] = false;
            minp[i * prime[j]] = prime[j];
            if (i % prime[j] == 0)
                break;
        }
    }
}
void invinit(int n){
    inv[1] = 1;
    for (int i = 2; i <= n;i++){
        //cout << mod - mod / i << "\n";
        inv[i] = (mod - mod / i) * inv[mod % i] % mod;
    }
}
int getnum(int p, int n){
    int ans = 0;
    while(n >= p){
        n /= p;
        ans++;
    }
    return ans;
}
ll mo(ll x){
    return (x % mod + mod) % mod;
}
void solve()
{
    int n;
    cin >> n;
    init(n);
    ll lcm = 1;
    for (int i = 1; i <= cnt;i++){
        int k = getnum(prime[i], n);
        for (int j = 0; j < k;j++)
            lcm = lcm * prime[i] % mod;
    }
    ll ans = 0;
    invinit(n);
    for (int i = 1; i <= n - 1;i++){
        ans = mo(ans + mo(lcm * mo((n - i) * (inv[i + 1]))));
    }
    //for (int i = 1; i <= n;i++)
        //cout << inv[i] << '\n';
    //cout << lcm;
    cout << ans;
    //cout<<getnum(2,9)<<'\n';
    //cout << getnum(3, 28);
}
int main()
{
    ios::sync_with_stdio(0), cin.tie(0), cout.tie(0);
    int t = 1;
    // cin >> t;
    while (t--)
        solve();
    return 0;
}

File: AStar/2024baiduStar/2024baiduStar01/B.py
isprime = [1] * 10000005
minp = [0] * 10000005
sum = [0] * 10000005
num = [0] * 10000005
f = [0] * 10000005
prime = [0] * 10000005
mod = 998244353
def init(n):
    global cnt
    cnt = 0
    isprime[1]= 0
    for i in range(2,n+1):
        if isprime[i]:
            cnt+=1
            prime[cnt]=i
            minp[i]=i
        j = 1
        while j <= cnt and i * prime[j] <= n:
            isprime[i*prime[j]]=False
            minp[i*prime[j]]=prime[j]
            if i % prime[j] == 0:
                break
            j+=1


lcm = 1
n = int(input())
init(10000000)
for i in range(2, n + 1):
    if isprime[i]:
        lcm = lcm * i
        num[i] += 1
        continue
    t = i
    mp = dict()
    while not isprime[t]:
        mp[minp[t]]+=1
        t /= minp[t]
    mp[t]+=1
    for c, d in mp:
        if num[c] < d:
            for z in range(0,d-num[c]):
                lcm = lcm * c
            num[c]=d
for i in range(n-1,0,-1):
    f[i] = lcm/i+1/i
ans = 0
for i in range(n-1,0,-1):
    sum[i] = f[i]*(n-i)+sum[i+1]
    ans=ans+sum[i]
print(ans % mod)
